pipeline/process_block.go

```diff
@@ -0,0 +1,135 @@
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/types/known/timestamppb"
+	"io"
+	"runtime/debug"
+)
+
+func (p *Pipeline) ProcessBlock(block *bstream.Block, obj interface{}) error {
+	p.reqCtx.StartSpan("process_block")
+	metrics.BlockBeginProcess.Inc()
+	clock := &pbsubstreams.Clock{
+		Number:    block.Num(),
+		Id:        block.Id,
+		Timestamp: timestamppb.New(block.Time()),
+	}
+	cursor := obj.(bstream.Cursorable).Cursor()
+	step := obj.(bstream.Stepable).Step()
+
+	p.reqCtx.SetAttributes(attribute.Int64("block_num", int64(block.Num())))
+
+	err := p.processBlock(block, clock, cursor, step)
+
+	if err != nil && err != io.EOF {
+		// TODO: this is weird should this be on the syccess case?
+		p.runPostJobHooks(clock)
+		p.reqCtx.SetErr(err)
+		return err
+	} else {
+		p.reqCtx.SetOk("")
+	}
+	return nil
+}
+
+func (p *Pipeline) processBlock(block *bstream.Block, clock *pbsubstreams.Clock, cursor *bstream.Cursor, step bstream.StepType) (err error) {
+	// TODO: should this move to the step new
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("panic at block %d: %s", block.Num(), r)
+			p.reqCtx.logger.Error("panic while process block", zap.Uint64("block_num", block.Num()), zap.Error(err))
+			p.reqCtx.logger.Error(string(debug.Stack()))
+		}
+	}()
+
+	switch {
+	case step.Matches(bstream.StepUndo):
+		return p.handleStepUndo(clock, cursor)
+
+	case step.Matches(bstream.StepStalled):
+		p.handleStepStalled(block.Num())
+		return
+
+	case step.Matches(bstream.StepIrreversible):
+		return p.handleStepIrreversible(block.AsRef())
+
+	case step.Matches(bstream.StepNew):
+		return p.handleStepNew(block, clock, cursor)
+	}
+
+	return nil
+}
+
+func (p *Pipeline) handleStepUndo(clock *pbsubstreams.Clock, cursor *bstream.Cursor) error {
+	p.reqCtx.AddEvent("handling_step_undo")
+	if err := p.forkHandler.handleUndo(clock, cursor, p.cachingEngine, p.storeAccessor, p.respFunc); err != nil {
+		return fmt.Errorf("reverting outputs: %w", err)
+	}
+	return nil
+}
+
+func (p *Pipeline) handleStepStalled(blockNum uint64) {
+	p.forkHandler.removeReversibleOutput(blockNum)
+	return
+}
+
+func (p *Pipeline) handleStepNew(block *bstream.Block, clock *pbsubstreams.Clock, cursor *bstream.Cursor) error {
+	execOutput, err := p.cachingEngine.NewExecOutput(p.blockType, block, clock, cursor)
+	if err != nil {
+		return fmt.Errorf("setting up exec output: %w", err)
+	}
+
+	if err := p.runPreBlockHooks(clock); err != nil {
+		return fmt.Errorf("pre block hook: %w", err)
+	}
+
+	// TODO: will this happen twice? blockstream also calls this at stopBluckNum
+	if err = p.FlushStores(block.Num()); err != nil {
+		return fmt.Errorf("failed to flush stores: %w", err)
+	}
+
+	if isStopBlockReached(clock.Number, p.reqCtx.StopBlockNum()) {
+		p.reqCtx.logger.Debug("about to save cache output",
+			zap.Uint64("clock", clock.Number),
+			zap.Uint64("stop_block", p.reqCtx.StopBlockNum()),
+		)
+		if err = p.cachingEngine.Flush(p.reqCtx.Context()); err != nil {
```

Comment:
could we haev `reqCtx` embed the `context.Context` ? So we could simply pass `reqCtx` where a `context.Context` is needed? It's an interface.